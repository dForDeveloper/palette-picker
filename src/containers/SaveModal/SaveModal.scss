.Modal {
  position: absolute;
  top: 50vh;
  margin-top: -240px;
  width: 320px;
  height: 240px;
  background-color: #ffffff;
  left: 50%;
  margin-left: -160px;
  z-index: 2;
  padding: 32px;
  border-radius: 8px;
}

.Modal--h4 {
  font-weight: 400;
  text-align: center;
  font-size: 1.2rem;
}

.Modal--input {
  outline: none;
  padding: 8px;
  width: 100%;
  margin-top: 16px;
  border-radius: 4px;
  border: 1px solid #777777;
  transition: all 0.2s ease;
  height: 36px;

  &:focus {
    border: 1px solid #0facf3;
  }
}

.Modal--button {
  outline: none;
  margin-top: 16px;
  width: 100%;
  height: 36px;
  padding: 6px;
  border-radius: 4px;
  border: none;
  background-color: #ffffff;
  cursor: pointer;
  background-color: #0facf3;
  color: #ffffff;
  transition: all 0.2s ease;

  &:hover {
    background-color: #10a5e8;
  }
}