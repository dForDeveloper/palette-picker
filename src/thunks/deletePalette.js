import { setError, toggleLoading, setModal, removePalette } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const deletePalette = (projectID, paletteID) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('DELETE');
      await fetchData(`/api/v1/palettes/${paletteID}`, options);
      dispatch(removePalette(projectID, paletteID));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error deleting palette: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}