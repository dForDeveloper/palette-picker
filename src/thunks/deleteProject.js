import { setError, toggleLoading, removeProject, setModal } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('DELETE');  
      await fetchData(`/api/v1/projects/${id}`, options);
      dispatch(removeProject(id));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error deleting project: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}