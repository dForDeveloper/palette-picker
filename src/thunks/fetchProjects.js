import { setProjects, toggleLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchProjects = () => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const projects = await fetchData('/api/v1/projects');
      dispatch(setProjects(projects));
    } catch (error) {
      dispatch(setError('Error fetch projects: ' + error.message));
    }
    dispatch(toggleLoading(false))
  }
}