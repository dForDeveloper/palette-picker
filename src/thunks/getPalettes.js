import { fetchData } from '../utils/api';
import { toggleLoading, setError, setPalettes } from '../actions';

export const getPalettes = (id) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const palettes = await fetchData(`/api/v1/projects/${id}/palettes`);
      dispatch(setPalettes(id, palettes));
    } catch (error) {
      dispatch(setError('Error getting palettes: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
};