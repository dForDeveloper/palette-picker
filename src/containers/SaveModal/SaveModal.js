import React, { Component } from 'react';
import { saveNewPalette } from '../../thunks/saveNewPalette';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

export class Modal extends Component {
  state = {
    projectName: '',
    paletteName: ''
  };

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    const { projectName, paletteName } = this.state;
    const { projects, colors, saveNewPalette } = this.props;
    await saveNewPalette(projectName, paletteName, projects, colors);
  };

  render() {
    return (
      <form className="Modal" onSubmit={this.handleSubmit}>
        <h4 className="Modal--h4">save palette</h4>
        <input
          className="Modal--input"
          name="projectName"
          placeholder="project name"
          value={this.state.projectName}
          onChange={this.handleChange}
          autoFocus={true}
          required 
        />
        <input
          className="Modal--input"
          name="paletteName"
          placeholder="palette name"
          value={this.state.paletteName}
          onChange={this.handleChange}
          required
        />
        <input type="submit" value="save" className="Modal--button" />
      </form>
    );
  }
}

export const mapStateToProps = (state) => ({
  colors: state.colors,
  projects: state.projects
});

export const mapDispatchToProps = (dispatch) => ({
  saveNewPalette: (projectName, paletteName, projects, colors) => {
    dispatch(saveNewPalette(projectName, paletteName, projects, colors));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(Modal);

Modal.propTypes = {
  colors: PropTypes.array,
  projects: PropTypes.array,
  saveNewPalette: PropTypes.func,
};