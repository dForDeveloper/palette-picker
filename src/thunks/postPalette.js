import { setError, setModal } from '../actions';
import { getProjects } from './getProjects';
import { getPalettes } from './getPalettes';
import { fetchData, createOptions } from '../utils/api';

export const postPalette = (projectID, name, colors) => {
  return async (dispatch) => {
    try {
      const project_id = parseInt(projectID);
      const [color1, color2, color3, color4, color5] = colors;
      const palette = {
        name,
        project_id,
        color1,
        color2,
        color3,
        color4,
        color5
      };
      const options = createOptions('POST', palette);
      await fetchData('/api/v1/palettes', options);
      await dispatch(getProjects());
      await dispatch(getPalettes(projectID));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error adding palette: ' + error.message));
    }
  }
}