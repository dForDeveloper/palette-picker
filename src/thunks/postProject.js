import { setError, addProject } from '../actions';
import { postPalette } from './postPalette';
import { fetchData, createOptions } from '../utils/api';

export const postProject = (projectName, paletteName, colors) => {
  return async (dispatch) => {
    try {
      const options = createOptions('POST', { name: projectName });
      const { id: projectID } = await fetchData('/api/v1/projects', options);
      dispatch(addProject({ id: projectID, name: projectName }));
      await dispatch(postPalette(projectID, paletteName, colors));
    } catch (error) {
      dispatch(setError('Error adding project: ' + error.message));
    }
  }
}