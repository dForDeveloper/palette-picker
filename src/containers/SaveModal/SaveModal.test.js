import React from 'react';
import { shallow } from 'enzyme';
import { postProject } from '../../thunks/postProject';
import { postPalette } from '../../thunks/postPalette';
import { SaveModal, mapStateToProps, mapDispatchToProps } from './SaveModal';

jest.mock('../../thunks/postProject');
jest.mock('../../thunks/postPalette');

const mockColors = ['#123456', '#234567', '#345678', '#456789', '#56789a'];

describe('SaveModal', () => {
  let wrapper;
  const mockProps = {
    colors: ['#aabbcc', '#bbccdd', '#ccddee', '#778899', '#ffffff'],
    projects: [
      { id: 1, name: 'my project' },
      { id: 2, name: 'other project' },
    ],
    postProject: jest.fn(),
    postPalette: jest.fn()
  };

  beforeEach(() => {
    wrapper = shallow(<SaveModal {...mockProps} />);
  });

  describe('render', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should call handleSubmit when the form is submitted', () => {
      jest.spyOn(wrapper.instance(), 'handleSubmit');
      wrapper.instance().forceUpdate();
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.find('.Modal').simulate('submit', mockEvent);
      expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
    });

    it('should call handleChange when the project input changes', () => {
      jest.spyOn(wrapper.instance(), 'handleChange');
      wrapper.instance().forceUpdate();
      const mockEvent = { target: { value: 'a'} };
      wrapper.find('[name="projectName"]').simulate('change', mockEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalled();
    });

    it('should call handleChange when the palette input changes', () => {
      jest.spyOn(wrapper.instance(), 'handleChange');
      wrapper.instance().forceUpdate();
      const mockEvent = { target: { value: 'b'} };
      wrapper.find('[name="paletteName"]').simulate('change', mockEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalled();
    });
  });

  describe('handleChange', () => {
    it('should set state with event.target.value', () => {
      const mockEvent = { target: { value: 'asdf' , name: 'projectName' } };
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('projectName')).toEqual('asdf');
    });
  });

  describe('handleSubmit', () => {
    it('should call postPalette if the project name exists', () => {
      wrapper.setState({ projectName: 'my project', paletteName: 'pal 1' });
      wrapper.instance().handleSubmit({ preventDefault: jest.fn() });
      expect(mockProps.postPalette).toHaveBeenCalled();
    });
    
    it('should call postProject if the project name is new', () => {
      wrapper.setState({ projectName: 'a', paletteName: 'b' });
      wrapper.instance().handleSubmit({ preventDefault: jest.fn() });
      expect(mockProps.postProject).toHaveBeenCalled();
    });
  });
});

describe('mapStateToProps', () => {
  it('should return a props object with the correct properties', () => {
    const mockState = {
      colors: mockColors,
      projects: [{ id: 1, name: 'my project' }],
      extraProperty: true
    };
    const expected = {
      colors: mockColors,
      projects: [{ id: 1, name: 'my project' }]
    };
    const result = mapStateToProps(mockState);
    expect(result).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn();
  const mappedProps = mapDispatchToProps(mockDispatch);

  it('should call postPalette', () => {
    const expected = postPalette(1, 'my palette', mockColors);
    mappedProps.postPalette(1, 'my palette', mockColors);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('should call postProject', () => {
    const expected = postProject('my project', 'my palette', mockColors);
    mappedProps.postProject('my project', 'my palette', mockColors);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});