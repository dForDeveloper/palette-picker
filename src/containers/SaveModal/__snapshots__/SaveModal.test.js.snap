// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveModal render should match the snapshot 1`] = `
<form
  className="Modal"
  onSubmit={[Function]}
>
  <h4
    className="Modal--h4"
  >
    save palette
  </h4>
  <input
    autoFocus={true}
    className="Modal--input"
    name="projectName"
    onChange={[Function]}
    placeholder="project name"
    required={true}
    value=""
  />
  <input
    className="Modal--input"
    name="paletteName"
    onChange={[Function]}
    placeholder="palette name"
    required={true}
    value=""
  />
  <input
    className="Modal--button"
    type="submit"
    value="save"
  />
</form>
`;
